test/jit/test_hooks.py

```diff
@@ -0,0 +1,357 @@
+import os
+import sys
+
+import torch
+from jit.test_hooks_modules import *
+
+# Make the helper files in test/ importable
+pytorch_test_dir = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
+sys.path.append(pytorch_test_dir)
+from torch.testing._internal.jit_utils import JitTestCase
+
+if __name__ == "__main__":
+    raise RuntimeError(
+        "This test file is not meant to be run directly, use:\n\n"
+        "\tpython test/test_jit.py TESTNAME\n\n"
+        "instead."
+    )
+
+
+# Tests for JIT forward hooks and pre-hooks
+class TestHooks(JitTestCase):
```

Comment:
One more test case I would add (if it's not there already) is accessing a submodule, storing it in a temporary variable, and then calling it:
```
m = torch.jit.script(module_with_hooks)
sm = m.submod_with_hooks
output = sm(input)
# check that output is ok, hooks should have been called for `sm`
```

Reply:
And then we also need a C++ version of this.